src/laser/Laser.cpp

```diff
@@ -35,70 +55,671 @@ Laser::InitData (const amrex::BoxArray& slice_ba,
+                }
+                const Complex lapA = lapR + I*lapI;
+                const Complex an00j00 = n00j00_arr(i,j,0) + I * n00j00_arr(i,j,1);
+                const Complex anp1jp1 = np1jp1_arr(i,j,0) + I * np1jp1_arr(i,j,1);
+                const Complex anp1jp2 = np1jp2_arr(i,j,0) + I * np1jp2_arr(i,j,1);
+                Complex rhs;
+                if (step == 0) {
+                    // First time step: non-centered push to go
+                    // from step 0 to step 1 without knowing -1.
+                    const Complex an00jp1 = n00jp1_arr(i,j,0) + I * n00jp1_arr(i,j,1);
+                    const Complex an00jp2 = n00jp2_arr(i,j,0) + I * n00jp2_arr(i,j,1);
+                    rhs =
+                        + 8._rt/(c*dt*dz)*(-anp1jp1+an00jp1)*edt1
+                        + 2._rt/(c*dt*dz)*(+anp1jp2-an00jp2)*edt2
+                        - 2._rt * chi_fac * isl_arr(i,j,chi) * an00j00
+                        - lapA
+                        + ( -6._rt/(c*dt*dz) + 4._rt*I*djn/(c*dt) + I*4._rt*k0/(c*dt) ) * an00j00;
+                } else {
+                    const Complex anm1jp1 = nm1jp1_arr(i,j,0) + I * nm1jp1_arr(i,j,1);
+                    const Complex anm1jp2 = nm1jp2_arr(i,j,0) + I * nm1jp2_arr(i,j,1);
+                    const Complex anm1j00 = nm1j00_arr(i,j,0) + I * nm1j00_arr(i,j,1);
+                    rhs =
+                        + 4._rt/(c*dt*dz)*(-anp1jp1+anm1jp1)*edt1
+                        + 1._rt/(c*dt*dz)*(+anp1jp2-anm1jp2)*edt2
+                        - 4._rt/(c*c*dt*dt)*an00j00
+                        - 2._rt * chi_fac * isl_arr(i,j,chi) * an00j00
+                        - lapA
+                        + ( -3._rt/(c*dt*dz) + 2._rt*I*djn/(c*dt) + 2._rt/(c*c*dt*dt) + I*2._rt*k0/(c*dt) ) * anm1j00;
+                }
+                rhs_arr(i,j,0) = rhs;
+                rhs_mg_arr(i,j,0) = rhs.real();
+                rhs_mg_arr(i,j,1) = rhs.imag();
+            });
+    }
+
+    // construct slice geometry
+    // Set the lo and hi of domain and probdomain in the z direction
+    amrex::RealBox tmp_probdom({AMREX_D_DECL(geom.ProbLo(Direction::x),
+                                             geom.ProbLo(Direction::y),
+                                             geom.ProbLo(Direction::z))},
+                               {AMREX_D_DECL(geom.ProbHi(Direction::x),
+                                             geom.ProbHi(Direction::y),
+                                             geom.ProbHi(Direction::z))});
+    amrex::Box tmp_dom = geom.Domain();
+    const amrex::Real hi = geom.ProbHi(Direction::z);
+    const amrex::Real lo = hi - geom.CellSize(2);
+    tmp_probdom.setLo(Direction::z, lo);
+    tmp_probdom.setHi(Direction::z, hi);
+    tmp_dom.setSmall(Direction::z, 0);
+    tmp_dom.setBig(Direction::z, 0);
+    amrex::Geometry slice_geom = amrex::Geometry(
+        tmp_dom, tmp_probdom, geom.Coord(), geom.isPeriodic());
+
+    slice_geom.setPeriodicity({0,0,0});
```

Comment:
does that mean that the MG solver does not have periodic BC, while the FFT solver has?

Reply:
Yes I assume so. The FFT solver could be improved to have better BC (Dirichlet or open BC).